libafl/src/inputs/mod.rs

```diff
@@ -117,15 +117,116 @@ impl Input for NopInput {
 }
 impl HasTargetBytes for NopInput {
     fn target_bytes(&self) -> OwnedSlice<u8> {
-        OwnedSlice::from(vec![0])
+        OwnedSlice::from(vec![])
+    }
+}
+
+impl HasLen for NopInput {
+    fn len(&self) -> usize {
+        0
+    }
+}
+
+/// Target bytes wrapper keeping track of the current read position.
+/// Convenient wrapper when bytes must be split it in multiple subinputs.
+#[derive(Debug)]
+pub struct BytesReader<'a, I>
+where
+    I: HasTargetBytes + HasLen,
+{
+    parent_input: &'a I,
+    pos: usize,
+}
+
+/// Representation of a partial slice
+/// This is used when providing a slice smaller than the expected one.
+/// It notably happens when trying to read the end of an input.
+#[derive(Debug)]
+pub enum PartialBytesSubInput<'a> {
+    /// The slice is empty, and thus not kept
+    Empty,
+    /// The slice is strictly smaller than the expected one.
+    Partial(BytesSubInput<'a>),
+}
+
+impl<'a> PartialBytesSubInput<'a> {
+    /// Consumes `PartialBytesSubInput` and returns true if it was empty, false otherwise.
+    #[must_use]
+    pub fn empty(self) -> bool {
+        matches!(self, PartialBytesSubInput::Empty)
+    }
+
+    /// Consumes `PartialBytesSubInput` and returns the partial slice if it was a partial slice, None otherwise.
+    #[must_use]
+    pub fn partial(self) -> Option<BytesSubInput<'a>> {
+        #[allow(clippy::match_wildcard_for_single_variants)]
+        match self {
+            PartialBytesSubInput::Partial(partial_slice) => Some(partial_slice),
+            _ => None,
+        }
+    }
+}
+
+impl<'a, I> BytesReader<'a, I>
+where
+    I: HasTargetBytes + HasLen,
```

Comment:
Why do we need TargetBytes here?

Reply:
To get the `OwnedSlice` from the input directly.
Another possibility is to take an `OwnedSlice<T>` as input to make this generic. Is it worth doing?

Reply:
Shouldn't this operate on `HasMutatorBytes` though?

Reply:
Also, I wonder if `HasTargetBytes` shouldn't actually return a "normal" slice, instead of an OwnedSlice?

Reply:
It's more restricting than necessary; HasTargetBytes is enough.

Reply:
HasMutatorBytes is more correct, HasTargetBytes means something different

Reply:
They happen to be the same for BytesInput but they are different for more abstract input types

Reply:
("the thing we send to the target" vs "the thing we mutate")

Reply:
Want to change it / any reason not to?

Reply:
sorry, didn't have time to check this week.
i was intending to use this in libafl_qemu, where we receive TargetBytes and wrap it whis this object to simplify the process of putting inputs in registers for example. doesn't it sound like TargetBytes-related? or maybe we need this as well for MutatorBytes?

Reply:
Ah.. The use case it not really mutator related then, right?
Might be somewhat independent (but maybe we can reuse some shared types then?)